### ISSUE TITLE
Improve workflow for Blue/Green Deployments that require a Migration

### ISSUE THREAD

**Issue by mmarcotte** [ustaxcourt/ef-cms#789]
In order to reduce confusion, possible human error, these are the proposed steps for performing a deployment that involves a Migration.

**Initiate**

This is currently performed manually.

- [ ] Determine and update the destination table name in the deploy table
- [ ] Change the `migration` flag to `true` in the deploy table

**Backup AWS Resources**

This is currently not performed. It may not be necessary. There will be a need to prune these backups. That should go in its own story if we maintain these backups

- [ ] Backup the destination table
- [ ] Backup Elasticsearch Index

**Destroy AWS Resources**

This is currently performed manually. Conceivably, this could be done as part of a workflow if the migrate flag is `true`

- [ ] Destroy the destination DDB table
- [ ] Destroy the destination ES cluster

**Recreate AWS Resources**

This is currently performed via the Web-Api Terraform deploy. 

- [ ] Create destination DDB table
- [ ] Create destination ES Cluster
- [ ] Create DDB Stream for Destination Table (Can the Lambda trigger be default disabled? This is because we don't want to trigger indexing while data is still being moved)

**Deploy Migration Infrastructure**

This is currently automated in the `migrate` step, but is manual for production because it does not currently have that `step` in the CircleCI Yaml. This requires the most recent branch to be pulled. It should probably be completely dockerized.

- [ ] Deploy Migration Infra (Lambda, SQS Queue, )
- [ ] Need to ready a new SQS queue to capture write events in Source Table to deliver to Source Table
- [ ] Need to link DDB Stream / Process Streams Lambda on Source Table / Current Color to populate this new SQS queue

**Setup Linkage between live application and Migration**

This is currently a little confusing, and I'm pretty sure is part of the previous step. We do make a lambda trigger to copy writes over from the Source table to the Destination table. However, they may be making it to the Destination table prematurely. We need to capture these writes into a queue, and then deliver them once the migration is complete. It's vital that we get this timing right.

- [ ] Enable a Stream/SQS queue to capture writes from Live table
- [ ] Create Take a snapshot/backup of the Live DDB table at a given timestamp
- [ ] Purge the Stream/SQS queue of writes so that it matches writes since the snapshot/backup was created
- [ ] Restore the snapshot/backup to a table (could this be the Destination Table or just the actual Source table?)

**Kickoff Migration**

This is currently automated in the `migrate` step, but is manual for production because it does not currently have that `step` in the CircleCI Yaml.

- [ ] Make sure the Streams Lambda trigger is `disabled` for the Destination Table
- [ ] Make sure the `refresh_interval` is `-1` for the Destination ES Cluster
- [ ] Kick off migration by sending table segments to migration SQS queue

**Wait for Migration to finish**

Circle attempts to do this, but it shouldn't for environments that house full size data. It will time out and just effectively do nothing while eating up circle credits. The lambdas that handle the migrations could provide feedback about progress, but that work will be covered [in another story](https://github.com/ustaxcourt/ef-cms/issues/952).

- [ ] Once the Lambdas cannot find anything in the Migration SQS queue for several minutes, (this depends on the Visibility timeout for the SQS queue right?) it should be done and kick off the next phase

**Were there any errors?**

So far, we haven't observed any errors. What happens when we do? In general, we would have to "stop the line" and figure out how to handle them. 

- [ ] Check the DLQ, check the error logs, figure out what to do: https://github.com/ustaxcourt/ef-cms/issues/953
- [ ] Do not proceed to the next step

**Destroy Migration Infrastructure**

Its job is done. Currently this is a manual step that needs to be performed. It is not automated on any environment. NOTE: this should not interfere with the queue/stream from the Live DDB table

- [ ] Run the existing Terraform script to destroy the migration queue and lambdas.

**Enable Lambda Trigger for Source Table SQS**

Now that we have migrated all of the information over from the Source table, it's time to turn on the hose from the Live DDB  writes that happened since we started the migration process. 

NOTE: these items need to be migrated, correct? 
NOTE ALSO: we need to figure out a reasonable threshold for when this is "empty" so that we can know that we are ready to switch colors

- [ ] Enable Lambda trigger to clear the queue/stream of writes from the Live DDB Table

**Enable the Lambda Trigger for the Destination Table**

The information should be fully copied into the destination table. It's now time to index it into ES. 

-[ ] Enable Lambda trigger to clear the queue/stream from the Destination Table

**Wait until queues are finished**

- [ ] Once queues are empty / caught up, it's time to test and/or switch colors

**Comment by matthopson** [comment]
Getting started with some draft A/C:

As THE ROBOTIC HUMAN ANALOG, I want to:
- Deploy and migrate via merge to deploying branch
- Avoid manually checking or writing to the deploy table (migrate flag, table version, etc...)
- Avoid manually setting environment variables specific to the deploy, migration, or color change
  - one exception may be the ENV variable (stg, prod, etc...)
- Avoid having to ensure you're on the correct / latest branch for the current deploy
- Avoid manually monitoring the status of the migration so the migration stack can be destroyed automatically
- Automatically destroy the migration infrastructure when the migration is complete
- Avoid manually monitoring the status of the Elasticsearch re-index so the verification and color switch steps can proceed automatically
- Automatically verify the ES indexing was successful and complete
- Automatically verify that there were no errors migrating to the new table
- Smoketest the system without mutating data
  - Log in as a user with creds from an environment variable
  - Use search features (case, practitioner, document, etc...)
  - Does NOT write to system tables or insert any test data (yet)
- Automatically switch colors when all verifications have completed
- Automatically back up and destroy the previous (source) table after the colors have switched
  - Back up east
  - Destroy east and west
- Automatically back up (snapshot) and destroy the previous ES cluster after the colors have switched
